### PR TITLE
Pair-wise analysis of frozen wheel offline chunks

### PR DESCRIPTION
**Pull request by josefbitzenhofer**
## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive analysis pipeline for neural ensemble reactivation, including new visualization and classification tools for place cells.
  - Added new data structure for representing sorted place cells and ensemble counts.
  - Implemented new utility functions for handling pre- and post-activation imaging epochs.
  - Enhanced spike preprocessing to include normalized spike estimates for improved analysis.

- **Refactor**
  - Modularized and clarified place cell detection and offline correlation analysis, with improved handling of different experimental conditions.
  - Unified and enhanced filtering logic for speed and position thresholds.
  - Centralized and strengthened validation checks in session correction workflows.
  - Improved data validation and consistency checks in trial processing.
  - Generalized consecutive true-value detection utilities for flexible thresholding.
  - Simplified trial info construction and improved cache path management.

- **Bug Fixes**
  - Fixed spelling in error message assertions for behavioral data tests.
  - Corrected minor behavioral logic in anticipatory licking detection.

- **Tests**
  - Expanded and refined test coverage for filtering logic, including new edge cases and combined filtering scenarios.
  - Updated test function names to reflect generalized consecutive true-value detection.

- **Chores**
  - Updated `.gitignore` to exclude `.npz` and `.pkl` files.
  - Updated dependency versions and cleaned up package requirements.
  - Added instructions for installing the OASIS submodule.
  - Added new debug configuration for ensemble reactivation analysis.
  - Renamed and relocated the OASIS submodule for improved project structure.
  - Updated session metadata constants.
  - Extended mouse genotype and sex classification lists.
  - Improved error reporting and logging in session processing.

### CODE REVIEW

File: viral/grosmark_analysis.py [hunk 1]
```diff
@@ -462,7 +557,7 @@ def circular_distance_matrix(activity_matrix: np.ndarray) -> np.ndarray:
     return circular_dist_matrix
 
 
-def binarise_spikes(spks: np.ndarray) -> np.ndarray:
+def binarise_spikes(spks: np.ndarray, mad_threshold: int = 1.5) -> np.ndarray:
```
Comment (neuromantic99) [inline viral/grosmark_analysis.py:560]:
```suggestion
def binarise_spikes(spks: np.ndarray, mad_threshold: float = 1.5) -> np.ndarray:
```

File: viral/grosmark_analysis.py [hunk 2]
```diff
@@ -45,12 +45,76 @@ def grosmark_place_field(
...
+            flu=spks, wheel_freeze=session.wheel_freeze
+        )
+        # TODO: should pre and post be binarised as one?
+        offline_spks_pre = binarise_spikes(
+            offline_spks_pre,
+            mad_threshold=1.25,
+        )
+        offline_spks_post = binarise_spikes(offline_spks_post, mad_threshold=1.25)
+        spks = np.hstack([offline_spks_pre, online_spks, offline_spks_post])
+        # TODO: shape is off by one frame (pre is one short), do we care?
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/grosmark_analysis.py:79]: I think when you've built the spks matrix, the final frame of the pre-wheel-freeze period is not included I don't think it will make a difference to analysis, but could be nice to include for sanity's sake

File: viral/grosmark_analysis.py [hunk 3]
```diff
@@ -45,12 +45,77 @@ def grosmark_place_field(
...
+            spks[
+                :,
+                session.wheel_freeze.pre_training_end_frame : session.wheel_freeze.post_training_start_frame,
+            ],
+            mad_threshold=1.5,
+        )
+        offline_spks_pre, offline_spks_post = get_preactivation_reactivation(
+            flu=spks, wheel_freeze=session.wheel_freeze
+        )
+        # TODO: should pre and post be binarised as one?
```
Comment (neuromantic99) [inline viral/grosmark_analysis.py:72]: very good point, it will affect the MAD. I'm not sure tbh :sweating:

File: viral/grosmark_analysis.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/grosmark_analysis.py:79]: mmm could this actually throw us off-by-one also when we're computing the frames that belong to a given trial?

File: viral/grosmark_analysis.py [hunk 4]
```diff
@@ -45,12 +45,77 @@ def grosmark_place_field(
...
+        offline_spks_pre = binarise_spikes(
+            offline_spks_pre,
+            mad_threshold=1.25,
+        )
+        offline_spks_post = binarise_spikes(offline_spks_post, mad_threshold=1.25)
+        spks = np.hstack([offline_spks_pre, online_spks, offline_spks_post])
```
Comment (neuromantic99) [inline viral/grosmark_analysis.py:78]: I would add an assertion to check this is the same size as the original spks matrix

File: viral/grosmark_analysis.py (hunk 1 above)
Comment (josefbitzenhofer) [reply viral/grosmark_analysis.py:558]: Shame on me !

File: viral/cache_2p_sessions.py [hunk 5]
```diff
@@ -27,8 +27,10 @@
 
 from viral.constants import (
     BEHAVIOUR_DATA_PATH,
+    CACHE_PATH,
     SPREADSHEET_ID,
     SYNC_FILE_PATH,
+    TEMP_CACHE_PATH,
```
Comment (josefbitzenhofer) [inline viral/cache_2p_sessions.py:33]: Please add to template

**Review by neuromantic99 — APPROVED** [review]

File: viral/constants_TEMPLATE.py [hunk 6]
```diff
@@ -11,6 +11,9 @@
 SYNC_FILE_PATH = SERVER_PATH / "DAQami"
 TIFF_UMBRELLA = SERVER_PATH / "2P"
 
+CACHE_PATH = ""
+TEMP_CACHE_PATH = ""
```
Comment (josefbitzenhofer) [inline viral/constants_TEMPLATE.py:15]: I love that

File: viral/imaging_utils.py [hunk 7]
```diff
@@ -112,6 +112,45 @@ def extract_TTL_chunks(
...
+            start, end = max(i - full_window, 0), i
+        else:
+            # central difference
+            start, end = i - half_window, i + half_window
+
+        displacement = positions[end] - positions[start]
+        duration = (end - start) / sampling_rate
+        speeds[i] = abs(displacement) / duration if duration > 0 else 0.0
+
+    return speeds
```
Comment (josefbitzenhofer) [inline viral/imaging_utils.py:151]: Very nice function, do you think, we will need a test?
